openvm/src/empirical_constraints.rs

```diff
@@ -0,0 +1,370 @@
+use itertools::Itertools;
+use openvm_circuit::arch::VmCircuitConfig;
+use openvm_sdk::StdIn;
+use openvm_stark_sdk::openvm_stark_backend::p3_field::PrimeField32;
+use powdr_autoprecompiles::empirical_constraints::{
+    intersect_partitions, DebugInfo, EmpiricalConstraints,
+};
+use powdr_autoprecompiles::DegreeBound;
+use std::collections::btree_map::Entry;
+use std::collections::HashMap;
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::trace_generation::do_with_trace;
+use crate::{CompiledProgram, OriginalCompiledProgram};
+
+struct Row {
+    pc: u32,
+    timestamp: (u32, u32),
+    cells: Vec<u32>,
+}
+
+/// Materialized execution trace, Indexed by time and by PC
+#[derive(Default)]
+struct Trace {
+    /// The raw rows, in any order
+    rows: Vec<Row>,
+}
+
+impl Trace {
+    fn rows_by_pc(&self) -> BTreeMap<u32, Vec<&Row>> {
+        self.rows.iter().fold(BTreeMap::new(), |mut acc, row| {
+            acc.entry(row.pc).or_insert(Vec::new()).push(row);
+            acc
+        })
+    }
+
+    fn rows_by_time(&self) -> Vec<&Row> {
+        self.rows
+            .iter()
+            .sorted_by_key(|row| row.timestamp)
+            .collect()
+    }
+}
+
+pub fn detect_empirical_constraints(
+    program: &OriginalCompiledProgram,
+    degree_bound: DegreeBound,
+    inputs: Vec<StdIn>,
```

Comment:
Note that the CLI only supports one input, but we can already do PGO on several inputs on Reth, so I added support for multiple inputs here too.